feat: Performance API: Transaction

brustolin opened the pull request:
## :scroll: Description

Created a folder with performance API.
Hacked the hell out of the SDK to be able to send a transaction to Sentry.

![image](https://user-images.githubusercontent.com/28265868/104965639-e4f6c280-59bd-11eb-9883-d32d0f4bb0c1.png)

![image](https://user-images.githubusercontent.com/28265868/104965723-1a031500-59be-11eb-9e1e-a438da0caa0e.png)

![image](https://user-images.githubusercontent.com/28265868/104965847-56367580-59be-11eb-940b-ba04b981bd95.png)

## :bulb: Motivation and Context

Add performance API to sentry-cocoa SDK.

## :green_heart: How did you test it?

This is a draft PR, trying it out from the Sentry iOS app, checking result in Sentry.
Test will came later.

## :pencil: Checklist

<!--- Put an `x` in the boxes that apply -->

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the CHANGELOG
- [ ] I updated the docs if needed
- [ ] Review from the native team if needed
- [x] No breaking changes

## :crystal_ball: Next steps

TBD

bruno-garcia commented on Samples/iOS-ObjectiveC/iOS-ObjectiveC/ViewController.m:
```diff
@@ -99,6 +100,19 @@ - (IBAction)captureException:(id)sender
+    fakeTransaction.timestamp = [start dateByAddingTimeInterval:2];
+    [fakeTransaction finish]; //use finish instead of setting timestamp
+    
+    [[SentrySDK.currentHub getClient] captureTransaction:fakeTransaction];
```
once we get `captureTransaction` to `SentryHub`, we can get rid of this, and on `[fakeTransaction finish]` we'll capture it to Sentry

bruno-garcia commented on Sources/Sentry/Transactions/ISentrySpan.h:
```diff
@@ -0,0 +1,17 @@
+//
+//  ISpan.h
+//  Sentry
+//
+//  Created by Dhiogo Brustolin on 05/01/21.
+//  Copyright © 2021 Sentry. All rights reserved.
+//
+
```
```suggestion
```

marandaneto replied:
this could be configured on Xcode, so it does not add any more.

bruno-garcia commented on Sources/Sentry/Transactions/SentrySpanContext.m:
```diff
@@ -0,0 +1,44 @@
+//
+//  SpanContext.m
+//  Sentry
+//
+//  Created by Dhiogo Brustolin on 05/01/21.
+//  Copyright © 2021 Sentry. All rights reserved.
+//
+
```
```suggestion
```

bruno-garcia commented on Sources/Sentry/Transactions/SentryTransaction.h:
```diff
@@ -0,0 +1,68 @@
+//
+//  SentryTransaction.h
+//  Sentry
+//
+//  Created by Dhiogo Brustolin on 11/01/21.
+//  Copyright © 2021 Sentry. All rights reserved.
+//
+
```
```suggestion
```

bruno-garcia commented:
We should agree on the next steps. I guess if we can `startTransaction` on the `Hub` and then have `finish` on the transaction itself capture it, that could be a good milestone. So build from here to do that, plus get tests in so we can merge.

Probably best to keep a branch to merge into, and we pull `master` into that branch from time to time. So that we can continue to release `master` without having to add the incomplete performance feature.

@philipphofmann @marandaneto wdyt?

bruno-garcia commented:
I changed the `base` branch to `feat/performance-monitoring`, which can be used for all PRs related to performance. Draft PR for that branch for visibility: https://github.com/getsentry/sentry-cocoa/pull/909

marandaneto commented:
> We should agree on the next steps. I guess if we can `startTransaction` on the `Hub` and then have `finish` on the transaction itself capture it, that could be a good milestone. So build from here to do that, plus get tests in so we can merge.
> 
> Probably best to keep a branch to merge into, and we pull `master` into that branch from time to time. So that we can continue to release `master` without having to add the incomplete performance feature.
> 
> @philipphofmann @marandaneto wdyt?

that would be ideal

philipphofmann commented on Sources/Sentry/Transactions/SentrySpanContext.h:
```diff
@@ -0,0 +1,37 @@
+
+- (instancetype)init;
+- (instancetype)initWithSampled:(BOOL)sampled;
+- (instancetype)initWithtraceId:(SentryId *)traceId
+                         spanId:(SentrySpanId *)spanID
```
```suggestion
- (instancetype)initWithTraceId:(SentryId *)traceId
                         spanId:(SentrySpanId *)spanId
```

philipphofmann commented on Sources/Sentry/Transactions/SentrySpanContext.m:
```diff
@@ -0,0 +1,59 @@
+
+- (instancetype)init
+{
+    return [self initWithtraceId:[[SentryId alloc] init]
+                          spanId:[[SentrySpanId alloc] init]
+                        parentId:nil
+                      andSampled:false];
```
`l`: We can simplify this a bit
```suggestion
    return [self initWithSampled:false];
```

philipphofmann commented on Sources/Sentry/Transactions/SentrySpanContext.m:
```diff
@@ -0,0 +1,59 @@
+- (NSMutableDictionary *)tags
+{
+    if (_tags == nil)
+        _tags = [[NSMutableDictionary alloc] init];
```
`l`: Why not initialize the dict in the constructor so we can remove the check for `nil` in here.

brustolin replied:
This part was copied from SentryEvent. I guess we could.

bruno-garcia replied:
Initializing in the constructor only we're certain this will always be used otherwise we're just allocating for no use.

philipphofmann replied:
I didn't think about that @bruno-garcia, so I'm fine with keeping it then.

philipphofmann commented on Sources/Sentry/Transactions/SentryTransaction.h:
```diff
@@ -0,0 +1,71 @@
+@property (nonatomic, strong)
+    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *_Nullable context;
+
+//-(instancetype)initWithName:(NSString*)name;
```
```suggestion
```

philipphofmann commented on Sources/Sentry/Transactions/SentrySpanContext.m:
```diff
@@ -0,0 +1,59 @@
+#import "SentryId.h"
+#import "SentrySpanId.h"
+
+@interface
+SentrySpanContext () {
+    NSMutableDictionary<NSString *, NSString *> *_tags;
+}
```
`l`: I think we could remove this and also `- (NSMutableDictionary *)tags`. We could just allocate the dictionary on line 47, or am I missing something?

philipphofmann commented on Sources/Sentry/Transactions/SentryTransaction.h:
```diff
@@ -0,0 +1,71 @@
+- (instancetype)initWithTransactionContext:(SentryTransactionContext *)context
+                                    andHub:(SentryHub *)hub;
+- (instancetype)initWithName:(NSString *)name
+                     context:(SentrySpanContext *)context
```
`l`:
```suggestion
                     spanContext:(SentrySpanContext *)spanContext
```

philipphofmann commented on Sources/Sentry/Transactions/SentryTransaction.h:
```diff
@@ -0,0 +1,71 @@
+    NSDictionary<NSString *, NSDictionary<NSString *, id> *> *_Nullable context;
+
+//-(instancetype)initWithName:(NSString*)name;
+- (instancetype)initWithTransactionContext:(SentryTransactionContext *)context
```
`l`:
```suggestion
- (instancetype)initWithTransactionContext:(SentryTransactionContext *)transactionContext
```

philipphofmann commented on Sources/Sentry/Transactions/SentryTransaction.m:
```diff
@@ -0,0 +1,118 @@
+        self.timestamp = [SentryCurrentDate date];
+    }
+
+    NSMutableDictionary *serializedData = @{
```
`l`:
```suggestion
    NSMutableDictionary<NSString *, id> *serializedData = @{
```

philipphofmann commented on Sources/Sentry/Transactions/SentryTransaction.m:
```diff
@@ -0,0 +1,118 @@
+    NSMutableDictionary *serializedData = @{
+        @"event_id" : self.eventId.sentryIdString,
+        @"timestamp" : [self.timestamp sentry_toIso8601String],
+        @"spans" : @[],
```
Why is this empty?

brustolin replied:
Bruno wanted to see some progress so I created this draft PR. Still work to do before consider it done.

philipphofmann commented on Sources/Sentry/Transactions/SentryTransaction.m:
```diff
@@ -0,0 +1,118 @@
+    return self;
+}
+
+/*-(instancetype)initWithName:(NSString*)name {
+   return self initWithName:name trace:trace andHub:
+    return ;
+}*/
```
```suggestion
```

philipphofmann commented on Sources/Sentry/Transactions/SentryTransaction.h:
```diff
@@ -0,0 +1,71 @@
+
+@class SentrySpanContext, SentryTransactionContext, SentryHub;
+
+@interface SentryTransaction : NSObject <SentrySerializable>
```
`h`: Please add `NS_SWIFT_NAME`

bruno-garcia replied:
What would the `NS_SWIFT_NAME` be? Just `Transaction`?
That's a very broad name, can we keep it SentryTransaction though?

brustolin replied:
I will use SentryTransaction as Java does.

philipphofmann reviewed (commented):
Very nice hacking. I'm excited to see some transactions in our SDK. 🥳 

I know this is a draft PR, but still, I made lots of comments, so we can get this PR ready for review.

Please move all public headers into the `Sources/Public` folder see https://github.com/getsentry/sentry-cocoa/blob/master/CONTRIBUTING.md#public-headers. Please also move all `.m` files to `Sources/Sentry`.

I'm looking forward to seeing some tests for all of this.

marandaneto commented on Sources/Sentry/Transactions/SentrySpanId.h:
```diff
@@ -0,0 +1,45 @@
+ * "12c2d058-d584-4270-9aa2-eca08bf20986". It is recommended to omit dashes and use UUID v4 in all
+ * cases.
+ */
+@interface SentrySpanId : NSObject
```
they are different, See java https://github.com/getsentry/sentry-java/blob/3c87a48442e536a41ed905d9e43bdc2f7b800176/sentry/src/main/java/io/sentry/SpanId.java#L20-L22
only 16 chars

philipphofmann replied:
Ah, thanks for pointing that out. The comment says 36 chars. It would be nice to update the comments then @brustolin 😀.

marandaneto commented:
@brustolin just to explain the comments prefixes:

`l`: low, that means, its not important for the reviewer, you could ignore but it'd be nice to be fixed.
`m`: mid, it's sort of nice that you both agree either on a solution or to ignore.
`h`; high, its really important for the reviewer, you definitely should look into it.

bruno-garcia commented on Sources/Sentry/Transactions/ISentrySpan.h:
```diff
@@ -0,0 +1,17 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol ISentrySpan <NSObject>
```
@philipphofmann what will we name the actual class? On Android we chose to use `I` as a convention for such cases where the interface matches the implementation name .. Actually I think we just endded up prefixing all with `I` at the end, similar to C#

philipphofmann replied:
I know the `I` pattern, and personally, I'm not a huge fan of it. I think you should avoid it whenever possible. Name the interface after the concept and the actual implementation should tell more about the implementation itself. For example, in Java, you have `List` and `ArrayList`, `CopyOnWriteArrayList`, ...
The `I` pattern is more common in Java and C#, but I don't think it is common in Objective-C or Swift, or am I wrong @brustolin?

marandaneto replied:
I guess they can have the same name, but the file has a different extension, using `@interface` and `@ implementation `, if this holds true, I'd avoid the `I` as well

brustolin replied:
Thats correct, the I pattern is not common in Objc. Since I was testing I just copied from java implementation until I realized I didn't need it to capture a transaction, so, some classes are not being used yet.

Open to new name options.

Apple naming suggestion: 
"Most protocols group related methods that aren’t associated with any class in particular. This type of protocol should be named so that the protocol won’t be confused with a class. A common convention is to use a gerund (“...ing”) form"

"Some protocols group a number of unrelated methods (rather than create several separate small protocols). These protocols tend to be associated with a class that is the principal expression of the protocol. In these cases, the convention is to give the protocol the same name as the class."

I guess we could go for the same name?

marandaneto replied:
sounds good to me

brustolin commented on Sources/Sentry/SentryEnvelope.m:
```diff
@@ -126,45 +126,11 @@ - (instancetype)initWithEvent:(SentryEvent *)event
 
     return [self
-        initWithHeader:[[SentryEnvelopeItemHeader alloc] initWithType:SentryEnvelopeItemTypeEvent
+        initWithHeader:[[SentryEnvelopeItemHeader alloc] initWithType:event.type
```
Wondering whether this change can cause some errors.

marandaneto replied:
@philipphofmann

brustolin commented:
I think I fixed everything you guys commented.

By Bruno request I changed SentryTransaction inheritance to SentryEvent and remove some of the previous SentryTransaction handlers like "sendTransaction" from SentryClient to use "sendEvent", this way the client can prepare the envelope with all required information.

SentryTransaction tests included

marandaneto commented on Sources/Sentry/Public/SentryClient.h:
```diff
@@ -41,6 +41,9 @@ SENTRY_NO_INIT
 - (SentryId *)captureEvent:(SentryEvent *)event
                  withScope:(SentryScope *)scope NS_SWIFT_NAME(capture(event:scope:));
 
+- (SentryId *)captureTransaction:(SentryTransaction *)transaction
```
missing the `withScope`

marandaneto replied:
well, every other method does, `captureEvent` requires a Scope, and right now `captureTransaction` is instantiating a new Scope every time, so either we take a Scope param or pass null, so we save the allocation if not using it at all.

bruno-garcia replied:
> well, every other method does,

It's not really a reason

> captureEvent requires a Scope, and right now captureTransaction is instantiating a new Scope every time,

This is a good point.

@brustolin Can we pass `nil` instead of creating a scope if we're not using it?

> so either we take a Scope param or pass null, so we save the allocation if not using it at all.

Yes, this ^ please

brustolin replied:
For now, it does not accept null in the parameter, and it uses the parameter in a ton of other calls. I guess I could change it all until the sendEvent, but I would have to add some extra flow to verify if scope is nil.

bruno-garcia replied:
OK lets merge this and follow up with that on a next PR.

marandaneto commented on Sources/Sentry/Public/SentryHub.h:
```diff
@@ -2,7 +2,7 @@
 
 @class SentryEvent, SentryClient, SentryScope, SentrySession, SentryUser, SentryBreadcrumb,
-    SentryId, SentryUserFeedback, SentryEnvelope;
+    SentryId, SentryUserFeedback, SentryEnvelope, SentryTransactionContext, SentryTransaction;
```
on Java/.NET, we call it only `TransactionContext`

brustolin replied:
This was one of the first things I added, was trying to figure Objc SDK naming convention. I guess I can fallow Java naming correct?

marandaneto replied:
mmm usually we just call as it is, we only add `Sentry` in front of what could bring import conflicts, so we run into issues like https://github.com/getsentry/sentry-dart/issues/250

yeah I'd prefer to keep rather the same whenever possible.

bruno-garcia replied:
In Objective-C it's even more important to avoid conflicts since it'll happen even outside of an `import` like in Dart. So I'd say prefixing with Sentry here is probably the best option

bruno-garcia replied:
We tested it and it indeed conflicted. So if someone adds an OpenTelemetry library later it would not compile. This is why we have pretty much everything here prefixed with `Sentry`. I'll resolve this with this reason, feel free to reopen if strong disagree.

marandaneto replied:
that makes sense, actually, we had a bug with KSCrash about that, good catch

marandaneto commented on Sources/Sentry/SentryClient.m:
```diff
@@ -231,6 +232,11 @@ - (SentryId *)sendEvent:(SentryEvent *)event
 
+- (SentryId *)captureTransaction:(SentryTransaction *)transaction
+{
+    return [self captureEvent:transaction withScope:[[SentryScope alloc] init]];
```
this could call either `captureEnvelope` or the transport directly, remember to append those values to the transaction like `release`, `environment`, `tags`, probably from SentryOptions

brustolin replied:
Thats the point. Capture event append all these values already, lets say the SDK grows and more properties are necessary, we would have to change in one place. Do you think the best option it is to make them part ways?

marandaneto replied:
at least on Java/Android we've decided to not do it because transactions are supposed to be cheap, if we do captureEvent on Android it runs event processors, beforeSend etc, reads available memory, free storage, etc, it'd be very expensive and also having a bunch of data in the JSON that are not that useful, so we decided going forward adding the missing attributes later on (based on needs), I believe on iOS would be the same.

marandaneto replied:
you could still extract a method with the common attributes from captureEvent and captureTransaction and call that method, right, from both places.

bruno-garcia replied:
Since Java can also be used on the server and have hundreds or thousands of transactions per second, that was critical there. In here I suggested we just extend Transaction (like it was done on @HazAT 's SDKs like JS) since it's a client based transaction and also since it's simpler to implement (splitting things have turned out to be a huge head ache on both Java and .NET even). Since no server side SDK here we have less reason to do it so to move faster I'd vote for inhertiting from Event to add the new fields.

Also, captureEvent is what's used on all the test infrastructure. So the captureEnvelope before was hard to test in this repo, but the addded benefit to get all the device context was an added bonus,

marandaneto replied:
> splitting things have turned out to be a huge head ache on both Java and .NET even

I'm not aware of that, why's that?

This will for sure generate a new source of issues here and/or Android, it was decided not to do it on Android to be cheap, either on payload or memory footprint, now we re doing the opposite here for no reason, Mobile has limited resources and will offer if used heavily, think about every transaction doing IO to read available storage, memory, etc... I'm against it till we decide to read only the min. and useful data to be sent over with transactions, your call.
On the Android side, issues will come, where is the device context? why do I have it on iOS and not on Android? and the explanation will be actually the opposite.

bruno-garcia replied:
> > splitting things have turned out to be a huge head ache on both Java and .NET even
> 
> I'm not aware of that, why's that?
> 

Without a clear guidence of what should be set to both transaction and events it happened on both .NET and Java that neither added critical data like environment, release, tags. A lot more code was put into splitting things, to deal with mispatches such as TransactionName vs Transaction (string), and it's a diviation from the current spec that is just unknown territory. This had startTimestamp already on the event since v5 so it was already trying to prepare to have Event match both Transaction and Event.

> This will for sure generate a new source of issues here and/or Android, it was decided not to do it on Android to be cheap, either on payload or memory footprint, now we re doing the opposite here for no reason, Mobile has limited resources and will suffer if used heavily, think about every transaction doing IO to read available storage, memory, etc... I'm against it till we decide to read only the min. and useful data to be sent over with transactions, your call.
> On the Android side, issues will come, where is the device context? why do I have it on iOS and not on Android? and the explanation will be actually the opposite.

>  Android to be cheap

It was decided on Spring to be cheap. Android will have a transaction every few seconds or minutes, compared to hundreds of transactions per second on Spring. So the reason is last important here IMHO

marandaneto commented on Sources/Sentry/Transactions/SentrySpanContext.h:
```diff
@@ -0,0 +1,37 @@
+
+@property (nonatomic, strong) SentryId *traceId;
+
+@property (nonatomic, strong) SentrySpanId *spanId;
+
+@property (nonatomic, strong) SentrySpanId *_Nullable parentSpanId;
```
we could call them `SpanId` only instead of `SentrySpanId`, Java/.NET is like this

bruno-garcia replied:
I believe, like the previous comment must keep the prefix `Sentry`. Another one I'll close but against feel free to reopen if strong disagree. I pinged @mitsuhiko to confirm on this but afaik these types need to be prefixed. @HazAT maybe can chime in

marandaneto commented on Sources/Sentry/Transactions/SentrySpanContext.h:
```diff
@@ -0,0 +1,37 @@
+
+@property (nonatomic, copy) NSString *_Nullable operation;
+
+@property (nonatomic, copy) NSString *_Nullable spanDescription;
```
why not only `description`

brustolin replied:
description is a Objc NSObject method already. Like toString from java/.NET. I guess we could override it and create the setDescription, I dont know if  I like this option.

marandaneto replied:
oh I see its a reserved word, well that's ok then I'd say, to keep as it is

marandaneto commented:
> I think I fixed everything you guys commented.
> 
> By Bruno request I changed SentryTransaction inheritance to SentryEvent and remove some of the previous SentryTransaction handlers like "sendTransaction" from SentryClient to use "sendEvent", this way the client can prepare the envelope with all required information.
> 
> SentryTransaction tests included

nice, thanks for that.

I've written a few more comments.
btw please resolve the comments that are already addressed, it's easier to review :)

brustolin commented on Sources/Sentry/SentryEnvelope.m:
```diff
@@ -130,6 +131,24 @@ - (instancetype)initWithEvent:(SentryEvent *)event
+
+    if (nil != error) {
+        transaction.context = nil;
+        transaction.sdk = nil;
```
Code removed to use send event instead.

brustolin commented on Sources/Sentry/SentryEnvelope.m:
```diff
@@ -130,6 +131,24 @@ - (instancetype)initWithEvent:(SentryEvent *)event
+        transaction.context = nil;
+        transaction.sdk = nil;
+        error = nil;
+        json = [SentrySerialization dataWithJSONObject:[transaction serialize] error:&error];
```
This code was removed. Using sendEvent instead

bruno-garcia reviewed (commented):
I see some tests already. Is this still draft?
What's missing to get merged other than address open review comments?

bruno-garcia commented on Sources/Sentry/Transactions/SentrySpanId.h:
```diff
@@ -0,0 +1,45 @@
+- (instancetype)initWithUUID:(NSUUID *)uuid;
+
+/**
+ * Creates a SentryId from a 32 character hexadecimal string without dashes such as
+ * "12c2d058d58442709aa2eca08bf20986" or a 36 character hexadecimal string such as such as
+ * "12c2d058-d584-4270-9aa2-eca08bf20986".
```
Also seems wrong

brustolin commented:
Well, now I think I have addressed all comments. Still figuring out this way of using GitHub.
A couple of discussion the be solved yet, but I think is ready to be merged, and let's solve these bigger issues in future PRs.

bruno-garcia reviewed (commented):
Lets just address the last open issues pending replies from folks in Vienna but other than those points, LGTM

bruno-garcia reviewed (approved):
@brustolin I believe we agreed on one point, pass `nil` as `scope`. Other than that I think we can merge for now. The Transaction/Event discussion we can address later at least, lets not block this PR on that
